File: JasSwift/Sources/Panels/BrushesPanel.swift
import Foundation

/// Brushes panel menu definition.
///
/// Source of truth is workspace/panels/brushes.yaml's `menu:` block; the
/// generic reader (`menuItemsFromYaml`) builds the items from the compiled
/// bundle. Several rows share one YAML action and so arrive param-folded
/// from the generic menu builder (`set_brush_view_mode:thumbnail`,
/// `set_brush_thumbnail_size:small`, `toggle_brush_category:calligraphic`,
/// `open_brush_options:create`); `dispatch` / `isCheckedWithModel` split the
/// value suffix back off and route through the shared YAML action pipeline,
/// matching the sibling Stroke / Swatches panels.
public enum BrushesPanel {
    public static func menuItems() -> [PanelMenuItem] {
        menuItemsFromYaml("brushes_panel_content")
    }

    public static func dispatch(_ cmd: String, addr: PanelAddr,
                                layout: inout WorkspaceLayout,
                                model: Model? = nil) {
        if cmd == "close_panel" { layoutApply(&layout, opClosePanel(addr)); return }
        guard let model = model else { return }
        // Param-folded radio rows: split "<action>:<value>" and run the
        // underlying YAML action with the declared param key, so both panel
        // state and any downstream effects update — matching the in-panel
        // controls and the sibling Stroke / Swatches dispatch.
        if let v = strip(cmd, prefix: "set_brush_view_mode:") {
            runYamlActionByName("set_brush_view_mode", params: ["view_mode": v], model: model)
            return
        }
        if let v = strip(cmd, prefix: "set_brush_thumbnail_size:") {
            runYamlActionByName("set_brush_thumbnail_size", params: ["size": v], model: model)
            return
        }
        if let v = strip(cmd, prefix: "toggle_brush_category:") {
            runYamlActionByName("toggle_brush_category", params: ["type": v], model: model)
            return
        }
        if let v = strip(cmd, prefix: "open_brush_options:") {
            runYamlActionByName("open_brush_options", params: ["mode": v], model: model)
            return
        }
        // "Open Brush Library" is a disabled placeholder (no YAML action),
        // so its command is empty — nothing to dispatch.
        guard !cmd.isEmpty else { return }
        runYamlActionByName(cmd, params: [:], model: model)
    }

    public static func isChecked(_ cmd: String, layout: WorkspaceLayout) -> Bool {
        false
    }

    /// Variant with the model in scope so the view-mode / thumbnail-size /
    /// category radios render their checkmark from the Brushes panel state,
    /// matching the YAML `checked_when` expressions.
    public static func isCheckedWithModel(_ cmd: String, model: Model?) -> Bool {
        guard let model = model else { return false }
        let pid = "brushes_panel_content"
        if let v = strip(cmd, prefix: "set_brush_view_mode:") {
            return (model.stateStore.getPanel(pid, "view_mode") as? String) == v
        }
        if let v = strip(cmd, prefix: "set_brush_thumbnail_size:") {
            return (model.stateStore.getPanel(pid, "thumbnail_size") as? String) == v
        }
        if let v = strip(cmd, prefix: "toggle_brush_category:") {
            if let arr = model.stateStore.getPanel(pid, "category_filter") as? [Any] {
                return arr.contains { ($0 as? String) == v }
            }
            return false
        }
        return false
    }

    private static func strip(_ s: String, prefix: String) -> String? {
        s.hasPrefix(prefix) ? String(s.dropFirst(prefix.count)) : nil
    }
}

File: jas_dioxus/src/panels/brushes_panel.rs
//! Brushes panel menu definition (BRUSHES.md).
//!
//! The panel body (library disclosure headers + brush tiles + footer
//! toolbar) is rendered by the generic YAML interpreter from
//! `workspace/panels/brushes.yaml`; this module provides the hamburger
//! menu wiring so the panel integrates with the Window menu and panel-menu
//! chrome.
//!
//! Brushes is a TOGGLE-ONLY panel: it is not part of the default layout.
//! `Window > Brushes` summons it on demand (see the `toggle_panel_brushes`
//! arm in `menu_bar.rs`, gated on group membership so the first click adds
//! the panel via `WorkspaceLayout::show_panel`).
//!
//! Each menu entry fires a YAML action (see workspace/actions.yaml); the
//! dispatch here routes the command string through the shared
//! `dispatch_action` pipeline (the same path the panel-body buttons use)
//! and handles `close_panel` locally.
//!
//! The menu's CHECK MARKS are NOT here. They are brushes.yaml's own
//! `checked_when:` predicates, evaluated by `panels::panel_is_checked`
//! through the shared menu-state evaluator. This module used to carry an
//! `is_checked` returning `false` whose comment said that was already true;
//! it was not, and all eleven check marks were dead as a result.

use crate::workspace::app_state::AppState;
use crate::workspace::workspace::PanelAddr;
use super::panel_menu::PanelMenuItem;

/// Menu items for the Brushes panel.
///
/// Source of truth is workspace/panels/brushes.yaml's `menu:` block; the
/// generic reader builds the items from the compiled bundle.
pub fn menu_items() -> Vec<PanelMenuItem> {
    super::panel_menu::menu_items_from_yaml("brushes_panel_content")
}

/// Dispatch a menu command for the Brushes panel.
///
/// `close_panel` is handled here. Every other command is a brush action
/// defined in actions.yaml and is routed through the shared
/// `dispatch_action` pipeline so the hamburger menu fires the same effect
/// chain as the panel-body wiring.
///
/// A radio row's command arrives FOLDED from the generic builder
/// (`set_brush_view_mode:list`), because several rows share one YAML action.
/// `panel_menu::action_and_params` unfolds it back into the action and the
/// entry's declared `params`, so `param.view_mode` resolves. Passing the folded
/// string through with an empty params map — which is what this did — named no
/// action at all, so every Brushes menu row was a silent no-op.
pub fn dispatch(cmd: &str, addr: PanelAddr, state: &mut AppState) {
    // The action evaluates with this panel's scope (selected_library,
    // selected_brushes), which the app scope alone does not carry.
    super::panel_menu::dispatch_yaml_menu("brushes_panel_content", cmd, addr, state);
}
